stdlib/json.gr

```diff
@@ -0,0 +1,2109 @@
+              }
+            }
+
+            if (format.objectFormat == OneObjectEntryPerLine) {
+              match (implHelper.printIndentation) {
+                Some(printIndentation) => printIndentation(elemLevel),
+                None => void,
+              }
+            }
+
+            let (key, value) = currentHead
+
+            implHelper.emitEscapedQuotedStringPretty(key)
+
+            Buffer.addChar(':', buffer)
+            match (format.objectFormat) {
+              CompactObjectEntries => void,
+              SpacedObjectEntries | OneObjectEntryPerLine => {
+                Buffer.addChar(' ', buffer)
+              },
+            }
+
+            match (printElementPretty(value, implHelper, elemLevel)) {
+              None => void,
+              err => return err,
+            }
+
+            match (currentRest) {
+              [] => break,
+              [newHead, ...newRest] => {
+                currentHead = newHead
+                currentRest = newRest
+              },
+            }
+          }
+
+          if (format.objectFormat == OneObjectEntryPerLine) {
+            match (implHelper.printNewLine) {
+              Some(printNewLine) => printNewLine(),
+              None => void,
+            }
+            match (implHelper.printIndentation) {
+              Some(printIndentation) => printIndentation(indentationLevel),
+              None => void,
+            }
+          }
+
+          Buffer.addChar('}', buffer)
+
+          return None
+        },
+      }
+    },
+  }
+}
+
+let isCodePointInBasicMultilingualPlane = (code: Number) =>
+  code >= 0x0000 &&
+    code <= 0xFFFF
+
+let isHighSurrogate = (code: Number) => code >= 0xD800 && code <= 0xDBFF
+
+let isLowSurrogate = (code: Number) => code >= 0xDC00 && code <= 0xDFFF
+
+let combineSurrogatePairToCodePoint =
+  (
+    highSurrogate: Number,
+    lowSurrogate: Number,
+  ) => {
+  // If this was a method exposed by itself in a library then it should check the
+  // ranges of the input surrogates, but here it's necessary because checks are made
+  // as part of the parsing logic.
+  ((highSurrogate - 0xD800) << 10) + (lowSurrogate - 0xDC00) + 0x10000
+}
+
+let makeJsonWriter = (format: FormattingSettings, buffer: Buffer.Buffer) => {
+  let printNewLine = match (format.lineEnding) {
+    NoLineEnding => None,
+    LineFeed =>
+      Some(() => {
+        Buffer.addChar('\n', buffer)
+      }),
+    CarriageReturnLineFeed =>
+      Some(() => {
+        Buffer.addChar('\r', buffer)
+        Buffer.addChar('\n', buffer)
+      }),
+    CarriageReturn =>
+      Some(() => {
+        Buffer.addChar('\r', buffer)
+      }),
+  }
+
+  let printIndentation = match (format.indentation) {
+    IndentWithTab =>
+      Some(indentationLevel => {
+        for (let mut count = 0; count < indentationLevel; count += 1) {
+          Buffer.addChar('\t', buffer)
+        }
+      }),
+    IndentWithSpaces(spacesPerIndentation) when spacesPerIndentation == 2 =>
```

Comment:
I'm assuming this is done to not allocate a closure? If so then maybe this should be done for 4-space indent as well since that's another popular alternative.

A comment indicating why this was done would be useful as well